### ISSUE TITLE
Augment Connections class

### ISSUE THREAD

**Issue by dmfigol** [nornir-automation/nornir#211]
This class could contain available connections as class attributes and instance attributes would be existing connections. 
Here is my previous proposal, which should be tested / discussed more:
```
class HostConnections(Dict[str, ConnectionPlugin]):
    available: Dict[str, Type[ConnectionPlugin]] = {}

    @classmethod
    def register(cls, name, value):
        cls.available[name] = value

    def open(self, name: str, ...):
        connection = self.available[name]()
        connection.open()
        self[name] = connection

    def close(self, ...): ...

    def close_all(self):
        for connection in self:
            self.close(connection)

    def get_or_create(self, ...): ...


class Host(object):
    ...
    def __init__(self, ...):
        ...
        self.connections = HostConnections()

    def open_connection(self, name: str, ...):
        self.connections.open(name, ...)
    ...
    
# can be called from plugin code or user code
HostConnections.register('netmiko', Netmiko)
HostConnections.register('napalm', Napalm)

```

**Comment by ktbyers** [comment]
@dmfigol Can you add a few sentences on what issue you are trying to solve here? How this helps you?

**Comment by dmfigol** [comment]
The main reason is to have an API to register connections (regardless if they are maintained by us or custom ones created by a user).
Currently we use the dictionary in `plugins.connections.__init__.py`:
```
available_connections: Dict[str, Type["ConnectionPlugin"]] = {
    "napalm": Napalm,
    "netmiko": Netmiko,
    "paramiko": Paramiko,
}
```
I would like to move it to the class `Connections` as a class attribute and have a `register` class method which will allow to do connections registration.

Second reason is to put all connection handling into that class, but this is more subjective.

**Comment by dbarrosop** [comment]
Pending implementation details I think it's a good approache. It might not be as easier as it sounds to implement due to circular dependencies but let's try it and see how it goes.

**Comment by dmfigol** [comment]
Implemented with #235
